### PR TITLE
[xray] Use pubsub instead of timeout for ObjectManager Pull.

### PR DESCRIPTION
**Pull request by elibol**
This PR removes the ObjectManager timeout for `Pull`. The ObjectDirectory interface method `Lookup` has been removed and replaced with methods `Subscribe` and `Unsubscribe`, which implement the existing GCS `Lookup` call followed by `RequestNotifications` whenever GCS `Lookup` fails to obtain object locations. This implementation eliminates the need to poll the GCS for object locations, which improves the performance of `Pull`.

### CODE REVIEW

File: src/ray/object_manager/object_directory.cc [hunk 1]
```diff
@@ -42,63 +67,77 @@ ray::Status ObjectDirectory::GetInformation(const ClientID &client_id,
...
-    existing_requests_[object_id] = ODCallbacks({success_callback, fail_callback});
-    status_code = ExecuteGetLocations(object_id);
-  } else {
-    // Do nothing. A request is in progress.
+ray::Status ObjectDirectory::SubscribeObjectLocations(const ObjectID &object_id,
+                                                      const OnLocationsFound &callback) {
+  ray::Status status = ray::Status::OK();
+  if (listeners_.find(object_id) != listeners_.end()) {
+    RAY_LOG(ERROR) << "Duplicate calls to SubscribeObjectLocations for " << object_id;
+    return ray::Status::Invalid("Cannot do things.");
```
Comment (robertnishihara) [inline src/ray/object_manager/object_directory.cc]: I don't understand this message.

File: src/ray/object_manager/object_directory.cc [hunk 2]
```diff
@@ -2,9 +2,34 @@
...
-ObjectDirectory::ObjectDirectory(std::shared_ptr<gcs::AsyncGcsClient> gcs_client) {
+ObjectDirectory::ObjectDirectory(std::shared_ptr<gcs::AsyncGcsClient> &gcs_client) {
   gcs_client_ = gcs_client;
-};
+}
+
+void ObjectDirectory::RegisterBackend() {
+  auto object_notification_callback = [this](gcs::AsyncGcsClient *client,
+                                             const ObjectID &object_id,
+                                             const std::vector<ObjectTableDataT> data) {
```
Comment (eric-jj) [inline src/ray/object_manager/object_directory.cc]: const std::vector<ObjectTableDataT>& data？

File: src/ray/object_manager/object_directory.cc [hunk 3]
```diff
@@ -2,9 +2,34 @@
...
+    // Objects are added to this map in SubscribeObjectLocations.
+    auto entry = listeners_.find(object_id);
+    // Do nothing for objects we are not listening for.
+    if (entry == listeners_.end()) {
+      return;
+    }
+    // Obtain reported client ids.
+    std::vector<ClientID> client_ids;
+    for (auto item : data) {
```
Comment (eric-jj) [inline src/ray/object_manager/object_directory.cc]: for (auto& item ：data)

File: src/ray/object_manager/object_directory.cc [hunk 4]
```diff
@@ -42,63 +67,77 @@ ray::Status ObjectDirectory::GetInformation(const ClientID &client_id,
...
+                                          const OnLocationsFound &success_callback,
+                                          const OnLocationsFailure &fail_callback) {
+  JobID job_id = JobID::nil();
+  ray::Status status = gcs_client_->object_table().Lookup(
+      job_id, object_id, [this, success_callback, fail_callback](
+                             gcs::AsyncGcsClient *client, const ObjectID &object_id,
+                             const std::vector<ObjectTableDataT> &location_entries) {
+        // Build the set of current locations based on the entries in the log.
+        std::unordered_set<ClientID> locations;
+        for (auto entry : location_entries) {
```
Comment (eric-jj) [inline src/ray/object_manager/object_directory.cc]: for (auto& entry : location_entries) {

File: src/ray/object_manager/object_directory.cc [hunk 5]
```diff
@@ -42,63 +67,77 @@ ray::Status ObjectDirectory::GetInformation(const ClientID &client_id,
...
-  ray::Status status = gcs_client_->object_table().Lookup(
-      job_id, object_id, [this](gcs::AsyncGcsClient *client, const ObjectID &object_id,
-                                const std::vector<ObjectTableDataT> &data) {
-        GetLocationsComplete(object_id, data);
+  listeners_.emplace(object_id, LocationListenerState(callback));
+  GetLocations(
+      object_id,
+      [this, callback](const std::vector<ray::ClientID> &v,
+                       const ray::ObjectID &object_id) {
+        if (listeners_.count(object_id) > 0) {
```
Comment (eric-jj) [inline src/ray/object_manager/object_directory.cc]: Is the GetLocations is a sync call or async call?
If it is a sync call, we don't need the check here. If it is an async call, it means the listeners_ can be changed during the line, we must add lock to it.

**Review by stephanie-wang — CHANGES_REQUESTED** [review]
In general, let's try to think carefully about the places where we return `ray::Status` and what it means if we return a status that is not OK. If there are methods where we should always return OK, I think it's better to just return void and add fatal checks for non-OK situations. If there are places where we should return a status, we should document in the method what it means if the status is not OK.

File: src/ray/object_manager/object_directory.cc [hunk 6]
```diff
@@ -42,63 +67,77 @@ ray::Status ObjectDirectory::GetInformation(const ClientID &client_id,
...
-  // Build the set of current locations based on the entries in the log.
-  std::unordered_set<ClientID> locations;
-  for (auto entry : location_entries) {
-    ClientID client_id = ClientID::from_binary(entry.manager);
-    if (!entry.is_eviction) {
-      locations.insert(client_id);
-    } else {
-      locations.erase(client_id);
-    }
+ray::Status ObjectDirectory::UnsubscribeObjectLocations(const ObjectID &object_id) {
```
[the remark this answers is not in the document]
Comment (elibol) [reply src/ray/object_manager/object_directory.cc]: This now returns status of `CancelNotifications`.

File: src/ray/object_manager/object_directory.cc [hunk 7]
```diff
@@ -2,9 +2,37 @@
...
+      if (!object_table_data.is_eviction) {
+        client_id_set.insert(client_id);
+      } else {
+        client_id_set.erase(client_id);
+      }
+    }
+    std::vector<ClientID> client_id_vec(client_id_set.begin(), client_id_set.end());
+    entry->second.locations_found_callback(client_id_vec, object_id);
+  };
+  gcs_client_->object_table().Subscribe(UniqueID::nil(),
```
Comment (stephanie-wang) [inline src/ray/object_manager/object_directory.cc]: There is an unused return status here. Either return it or `RAY_CHECK_OK`.

File: src/ray/object_manager/object_directory.cc [hunk 8]
```diff
@@ -2,9 +2,37 @@
...
+    for (auto &object_table_data : data) {
+      ClientID client_id = ClientID::from_binary(object_table_data.manager);
+      if (!object_table_data.is_eviction) {
+        client_id_set.insert(client_id);
+      } else {
+        client_id_set.erase(client_id);
+      }
+    }
+    std::vector<ClientID> client_id_vec(client_id_set.begin(), client_id_set.end());
+    entry->second.locations_found_callback(client_id_vec, object_id);
```
Comment (stephanie-wang) [inline src/ray/object_manager/object_directory.cc]: Do you want to call the callback if `client_id_vec` is empty?

**Review by stephanie-wang — APPROVED** [review]

**Comment by elibol** [comment]
Tests look good: https://travis-ci.com/robertnishihara/ray-private-travis/builds/73884240

**Pull request merged**